src/solver/algorithms/norvig_solver.py

```diff
@@ -0,0 +1,145 @@
+        self.units = dict((square, [unit for unit in self.unit_list if square in unit])
+                        for square in self.squares)
+        self.peers = dict((square, set(sum(self.units[square],[]))-set([square]))
+                        for square in self.squares)
+
+    def cross(self, list_A, list_B):
+        """Cross product of elements in A and elements in B.
+        Keyword arguments:
+        list_A -- list with elements to be combined
+        list_B -- list with elements to be combined
+        """
+        return [a+b for a in list_A for b in list_B]
+
+    def parse_grid(self, grid):
+        """Converts grid to a dictionary of possible values, {square: digits}, or
+        return False if a contradiction is detected.
+        Keyword arguments:
+        grid -- it is a list containing the grid values, it will have a zero ('0') for
+        non-set values (i.e. ['0', '0', '3', '0', '2', '0', '6', '0', '0', '9',...)
+        """
+        ## To start, every square can be any digit; then assign values from the grid.
+        values = dict((square, self.digits) for square in self.squares)
+        for square, digit in self.grid_values(grid).items():
+            if digit in self.digits and not self.assign(values, square, digit):
+                return False
+        return values
+
+    def grid_values(self, grid):
+        """Converts grid into a dict of {square: char} with '0' for empties.
+        Returns the dictionary generated (i.e. OrderedDict([('A1', '2'), ('A2', '0')...).
+        Keyword arguments:
+        grid -- it is a list containing the grid values, it will have a zero ('0') for
+        non-set values (i.e. ['0', '0', '3', '0', '2', '0', '6', '0', '0', '9',...)"""
+        assert len(grid) == 81
+        grid_values_dict = self.order_dictionary(dict(zip(self.squares, grid)))
+        return grid_values_dict
+
+    def order_dictionary(self, dict):
+        """Orders a dictionary and returns an OrderedDict object with the result
+        Keyword arguments:
+        dict -- it is the dictionary to be ordered (i.e. {'I6': '7', 'H9': '9',...)"""
+        ordered_dict = collections.OrderedDict(sorted(dict.items()))
+        return ordered_dict
+
+    def assign(self, values, square, digit):
+        """Eliminates all the other values (except digit) from values[square] and propagate.
+        Returns values, except return False if a contradiction is detected.
+        Keyword arguments:
+        values -- dictionary containing the game values (i.e. {'I6': '7', 'H9': '9',...)
+        square -- string containing the square to be handled in the method (i.e. 'I6')
+        digit -- string containing the digit to be handled in the method (i.e. '7')
+        """
+        other_values = values[square].replace(digit, '')
+        if all(self.eliminate(values, square, digit2) for digit2 in other_values):
+            return values
+        else:
+            return False
```

Comment:
it is a good practice to be consistent with the return values, if you return boleans you should do it in all the cases if you return an array with values for the other case you should return an empty array or None